trie/statistics/trieStatistics.go

```diff
@@ -73,55 +91,125 @@ func (ts *trieStatistics) collectNodeStatistics(level int, size uint64, nodeStat
-		TotalNodesSize:     totalNodesSize,
-		TotalNumNodes:      totalNumNodes,
-		MaxTrieDepth:       ts.maxTrieDepth,
-		BranchNodesSize:    ts.branchNodes.nodesSize,
-		NumBranchNodes:     ts.branchNodes.numNodes,
-		ExtensionNodesSize: ts.extensionNodes.nodesSize,
-		NumExtensionNodes:  ts.extensionNodes.numNodes,
-		LeafNodesSize:      ts.leafNodes.nodesSize,
-		NumLeafNodes:       ts.leafNodes.numNodes,
-	}
+// GetTotalNodesSize will return the total size of all nodes
+func (ts *trieStatistics) GetTotalNodesSize() uint64 {
+	ts.mutex.RLock()
+	defer ts.mutex.RUnlock()
+
+	return ts.getTotalNodesSize()
+}
+
+func (ts *trieStatistics) getTotalNodesSize() uint64 {
+	return ts.branchNodes.nodesSize + ts.extensionNodes.nodesSize + ts.leafNodes.nodesSize
+}
+
+// GetTotalNumNodes will return the total number of nodes
+func (ts *trieStatistics) GetTotalNumNodes() uint64 {
+	ts.mutex.RLock()
+	defer ts.mutex.RUnlock()
+
+	return ts.getTotalNumNodes()
+}
+
+func (ts *trieStatistics) getTotalNumNodes() uint64 {
+	return ts.branchNodes.numNodes + ts.extensionNodes.numNodes + ts.leafNodes.numNodes
+}
+
+// GetMaxTrieDepth will return the maximum trie depth
+func (ts *trieStatistics) GetMaxTrieDepth() uint32 {
+	ts.mutex.RLock()
+	defer ts.mutex.RUnlock()
+
+	return ts.maxTrieDepth
+}
+
+// GetBranchNodesSize will return the size of all branch nodes
+func (ts *trieStatistics) GetBranchNodesSize() uint64 {
+	ts.mutex.RLock()
+	defer ts.mutex.RUnlock()
+
+	return ts.branchNodes.nodesSize
+}
+
+// GetNumBranchNodes will return the number of branch nodes
+func (ts *trieStatistics) GetNumBranchNodes() uint64 {
+	ts.mutex.RLock()
+	defer ts.mutex.RUnlock()
+
+	return ts.branchNodes.numNodes
+}
+
+// GetExtensionNodesSize will return the size of all extension nodes
+func (ts *trieStatistics) GetExtensionNodesSize() uint64 {
+	ts.mutex.RLock()
+	defer ts.mutex.RUnlock()
+
+	return ts.extensionNodes.nodesSize
+}
+
+// GetNumExtensionNodes will return the number of extension nodes
+func (ts *trieStatistics) GetNumExtensionNodes() uint64 {
+	ts.mutex.RLock()
+	defer ts.mutex.RUnlock()
+
+	return ts.extensionNodes.numNodes
+}
+
+// GetLeafNodesSize will return the size of all leaf nodes
+func (ts *trieStatistics) GetLeafNodesSize() uint64 {
+	ts.mutex.RLock()
+	defer ts.mutex.RUnlock()
+
+	return ts.leafNodes.nodesSize
+}
+
+// GetNumLeafNodes will return the number of leaf nodes
+func (ts *trieStatistics) GetNumLeafNodes() uint64 {
+	ts.mutex.RLock()
+	defer ts.mutex.RUnlock()
+
+	return ts.leafNodes.numNodes
 }
 
-// TrieStatsDTO holds the statistics for the trie
-type TrieStatsDTO struct {
-	Address        string
-	RootHash       []byte
-	TotalNodesSize uint64
-	TotalNumNodes  uint64
-	MaxTrieDepth   uint32
+// GetLeavesMigrationStats will return the leaves migration statistics
+func (ts *trieStatistics) GetLeavesMigrationStats() map[core.TrieNodeVersion]uint64 {
+	ts.mutex.RLock()
+	defer ts.mutex.RUnlock()
+
+	return ts.migrationStats
```

Comment:
we need to return a new map here. Otherwise, the mutex protection is useless.

Reply:
Refactored.